### PR TITLE
JBMC: Remove-returns per function

### PR DESCRIPTION
**Pull request by smowton**
Another day, another pass converted to per-function operation. This one is really quite awkward, though: remove-returns transforms `x = f();` into `f(); x = f#return_value;`, and doing this function-by-function with other passes interleaved necessarily introduces some discord when a caller that hasn't been converted deals with a callee that has, or vice versa.

I've had a stab at a least-bad solution, and come up with:

* When a given function is being remove-return'd, alter its own return statements and its callsites to expect remove-return style, creating `#return_value` variables on demand.
* Do *not* alter the function's type, in the symbol table or the goto function map, so that other functions that must create callsites treat it as if it had its "real" return type. For example, the string solver will special-case functions that it expects to return String-related types, and become confused if they have been voided.
* Any passes that run *after* remove-returns risk becoming confused as the function's type indicates it returns conventionally, but in fact it routes its return values via `#return_value`. However such a pass would normally assume all functions are void-typed in any case.

In summary, remove-returns is a really messy pass. The "proper" solution would be to separate the notions of return type and result type, and always use the latter plus a "get result" function that hides whether a function is in conventional or via-global style. However I suspect attempting to implement that now would be a fool's errand.

### CODE REVIEW

**Comment by smowton** [comment]
Test-gen bump: https://github.com/diffblue/test-gen/pull/1364

**Comment by tautschnig** [comment]
Could you please clarify what the first commit is supposed to achieve on its own? Only the second commit has a detailed commit message, yet is somewhat confusing as it heavily touches `remove_returns` once again.

**Comment by tautschnig** [comment]
> Do _not_ alter the function's type, in the symbol table or the goto function map, so that other functions that must create callsites treat it as if it had its "real" return type. For example, the string solver will special-case functions that it expects to return String-related types, and become confused if they have been voided.

I don't yet buy this one: my approach would have been to alter the type of a function `f` once done with all rewriting in `f`. There would of course be inconsistencies if not all available bodies in a `goto_functionst` had been converted, but that should be a transient state (and I don't think that's the one you are concerned about here).

If the string solver expects some functions to always have non-void return types, then the string solver needs to be fixed -- it should be able to handle the two cases (1) has a non-void return type and (2) has void type -> look at `x#return_value`.

Surely I'm missing something, so why is it that the string solver can't be fixed?

**Comment by smowton** [comment]
@tautschnig fixed the commits; the first one was supposed to do the real work, and the second one activate it in jbmc.

The string solver can be fixed, but my concern is any frontend pass that happens to examine the type of another function, e.g. to generate a call sequence, will need to be taught about `remove_returns` too. I think it will be better for the front-end to be able to ignore the existence of remove-returns (even though it would find out about it if it inspected the other function's body) and have to possibly teach passes that run *after* remove-returns to assume functions are void-typed, rather than the reverse.

I can also run a whole-program-level cleanup pass that makes the type void so everything is consistent *then* -- thus only passes that run during the loading process will need to care about the distinction.

**Comment by tautschnig** [comment]
> [...] frontend pass that happens to examine the type of another function, e.g. to generate a call sequence, [...]

Would you have a concrete example of such a pass? The notion of "frontend" isn't perfectly clear to me here (I would not consider any analysis running on goto programs a "frontend" type pass anymore as, to me, the transition from `codet` to `goto_programt` is the step where we leave the front end).

**Comment by smowton** [comment]
The string solver is the only one I know about so far. The plan here is for symex to load functions on demand, so in practice, briefly defining "front-end" as anything that sees the code before loading is finished, the front-end will constitute everything from java-bytecode-convert-method through to the solver. Everything in `java_bytecode_languaget` and `goto_convert` is currently expecting conventional returns, and everything after `process_goto_functions` is currently expecting remove-returns style, or can tolerate both styles, so it's a question of which batch you'd rather fix. I favoured the later stages because things that expect remove-returns are very likely already ignoring the return type declared in the symbol table since it's meaningless, whereas earlier stages can reasonably expect the symbol table to truthfully reflect the return type.

**Comment by tautschnig** [comment]
Just had some out-of-band sync, @smowton will post updates.

**Comment by smowton** [comment]
Going to try the approach where we *do* alter the types in remove-returns just as in the whole-program case, but teach the Java frontend and related passes not to assume the symbol table is accurate, but rather to always use `get_original_return_type` or some close relative thereof that patches over the differences between conventional and removed-return style.

**Comment by smowton** [comment]
@tautschnig updated this to your design idea; in fact it turned out the string solver doesn't inspect other functions' types; rather it was failing because string-related functions can initially be created as stubs, and then later provided with a body. I worked around this by having `java_bytecode_convert_method` check whether that is the case before annotating a stub, also providing a hook for driver programs that supply their own stubs. If there's demand that can be generalised to allow the driver to supply a lambda, but I expect the manner in which stub bodies are supplied will be amended before then anyhow, as it would simplify lazy loading if they were "right first time", rather than being post-hoc fixed up.

Long story short, remove-returns now edits types just like the whole-program pass, and it appears that the frontend components are not perturbed, at least in the tests supplied with jbmc and deeptest. Please re-review.

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp]: NOLINT needed

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp]: NOLINT needed

File: src/java_bytecode/java_entry_point.cpp
Comment (tautschnig) [inline src/java_bytecode/java_entry_point.cpp]: Is this necessary? The annotation you could use is marking it `extern`, which will avoid such initialisation in the C front-end. That might work for Java as well, because Java would not be using `is_extern` natively, I believe?

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Use `true` instead of `ID_1`. The right thing will happen.

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp]: Is this comment still accurate?

**Review by tautschnig — COMMENTED** [review]
I think this is ok, but it's too late to call this a serious review.

File: src/java_bytecode/java_entry_point.cpp
Comment (smowton) [reply src/java_bytecode/java_entry_point.cpp]: Disinclined to abuse the flag so, since I *am* providing a definition of the symbol, it just doesn't need start-of-day initialisation because we know it is always written before reading

**Comment by smowton** [comment]
@tautschnig comments applied except for `is_extern`, which I am disinclined to use since it would misrepresent the linkage status of those variables.

**Comment by smowton** [comment]
@peterschrammel I suspect this is one you ought to review? Or possibly @cesaro?

File: src/goto-programs/remove_returns.cpp
Comment (peterschrammel) [inline src/goto-programs/remove_returns.cpp:81]: Ideally, we should have a consistency check function that could be used to add a debug-mode invariant whether a function has been remove_return'd or not. (Not TBD in  this PR, but a note for future hardening of goto passes.)

File: src/goto-programs/remove_returns.cpp
Comment (peterschrammel) [inline src/goto-programs/remove_returns.cpp]: use `i_it->make_assignment(assignment)`

File: src/java_bytecode/java_bytecode_convert_method_class.h
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method_class.h]: Is there a use case for this?

Comment (smowton) [reply src/java_bytecode/java_bytecode_convert_method_class.h]: Test-gen acts this way: stub functions are given a body shortly after creation.

**Comment by smowton** [comment]
@peterschrammel done one of your changes, replied to the other.

File: src/java_bytecode/java_bytecode_convert_method_class.h
Comment (peterschrammel) [reply src/java_bytecode/java_bytecode_convert_method_class.h]: Trying to understand how this will fit with the discussions around https://github.com/diffblue/cbmc/pull/1585 in the end...

Comment (smowton) [reply src/java_bytecode/java_bytecode_convert_method_class.h]: At the moment if you (the driver program) have a non-default plan for stubs, you should set this parameter. For example, test-gen derives from `java_bytecode_languaget` in order to make similar customisations, and will set this once this change hits CBMC.

Later I would like for stubbing to take place via a standard callback provided by languaget, such that the driver program supplies an implementation there instead of customising the language frontend. There could be standard implementations (replace with assume-false, add a Java-style nondet-object-graph stub, ...), or the driver could do something cleverer.

File: src/java_bytecode/java_entry_point.cpp
Comment (tautschnig) [reply src/java_bytecode/java_entry_point.cpp]: It just strikes me as something very magical, and very much of unclear value. I accept that it may not be necessary to do the initialisation, but is it really that much harm?

File: src/java_bytecode/java_bytecode_convert_method_class.h
Comment (tautschnig) [reply src/java_bytecode/java_bytecode_convert_method_class.h]: I have no idea about urgencies, but really I'd want to see this addressed before merging this PR (knowing that this blocks a series of other PRs). My problem with the chosen approach is that `function_is_stub` is a piece of configuration information, yet is never documented as such in, e.g., the symbol table. That is, the semantics of the program depend on `function_is_stub`, but there is no way to recover from one choice or the other.

There is an ugly option of delaying things until symbolic execution, which would figure out that no function definition is present, and thus create an assignment to the return-value variable instead (where the communication that a return value is required is implicit via the existence of the return-value symbol). The advantage is that goto programs can be changed up to the very last moment.

The proposed "standard callback" solution seems preferable, but is possibly more work. Indeed the discussion in #1585 should be driven to a conclusion.

File: src/java_bytecode/java_entry_point.cpp
Comment (smowton) [reply src/java_bytecode/java_entry_point.cpp]: At the moment at least with a largeish JAR available you'd get thousands of lines of `__CPROVER_initialize` that you didn't need, and therefore thousands of lines of state in goto-symex. Still, would it make you happier to invent a special `value` (`no_initialisation_required_exprt`?) that makes this clearer?

Comment (tautschnig) [reply src/java_bytecode/java_entry_point.cpp]: Ok, I get that this may yield many _avoidable_ instructions. What makes me unhappy is this special casing and dependency of otherwise completely unrelated bits of code.

What about renaming the `is_extern` flag and making it no longer C-specific? It could be `do_not_initialize`?

Comment (smowton) [reply src/java_bytecode/java_entry_point.cpp]: Don't want to rename it, but I'll gladly add an extra flag.

Comment (tautschnig) [reply src/java_bytecode/java_entry_point.cpp]: That would impact (de)serialisation of goto binaries, so take extra care!

Comment (smowton) [reply src/java_bytecode/java_entry_point.cpp]: Argh, then I won't. A comment on the type seems like the least invasive solution in that case.

Comment (tautschnig) [reply src/java_bytecode/java_entry_point.cpp]: You might just want to tell me to shut up and go with your existing solution plus a TODO saying that a different approach would fall into the class of #1148 :-)

File: src/java_bytecode/java_bytecode_convert_method_class.h
Comment (smowton) [reply src/java_bytecode/java_bytecode_convert_method_class.h]: So this actually isn't quite the same as the please-provide-a-stub callback, rather the question is, of a function that has *not yet* been loaded, will a body ever be provided? Therefore I propose for this PR I'll implement `function_will_have_body`, a callback provided by the driver program (jbmc will check `language_filest::lazy_method_map` to determine this, test-gen will always answer yes). That will remove the need for annotation on stub creation as is happening in this PR.

The provide-stub-body callback is an entirely separate matter, and should be pursued independently.

File: src/java_bytecode/java_entry_point.cpp
Comment (smowton) [reply src/java_bytecode/java_entry_point.cpp]: Nah that's okay, a type-comment (AKA an annotation, but that's an extra symbol field (which *is* a #1148 matter)) is strictly nicer than what I have now. I'll make a new version of this shortly, as the is-a-stub-callsite thing should be solved better anyhow.

**Comment by smowton** [comment]
@tautschnig made both these updates:

* Can-produce-function is now a property of `lazy_goto_modelt` that is routed through to `remove_returns`, so it can do the right thing without needing the frontend to correctly guess whether a stub will remain or be populated later on
* `#remove_return` variables that don't need initialisation now communicate this fact with an annotation (comment tagged onto their type) rather than the frontends having to encode that knowledge.

**Comment by smowton** [comment]
@tautschnig @peterschrammel  Travis failure is only on Mac builds cancelled due to problems at their end; this is ready for re-review.

File: src/goto-programs/remove_returns.cpp
Comment (peterschrammel) [inline src/goto-programs/remove_returns.cpp]: There's a make_assignment variant that takes a codet.

**Review by peterschrammel — COMMENTED** [review]
Looks good apart from some methods not having doxygen yet.

**Comment by tautschnig** [comment]
I have restarted the Mac builds: https://www.traviscistatus.com/ suggests that even the Mac side of the house should be back.

File: src/goto-programs/lazy_goto_functions_map.h
Comment (tautschnig) [inline src/goto-programs/lazy_goto_functions_map.h]: Needs documentation.

File: src/goto-programs/lazy_goto_model.h
Comment (tautschnig) [inline src/goto-programs/lazy_goto_model.h]: Interface to ... what?

File: src/goto-programs/lazy_goto_model.h
Comment (tautschnig) [inline src/goto-programs/lazy_goto_model.h:31]: I don't know what a `can_produce_functiont` is supposed to be, hence I can't judge whether this is-a relation is justified.

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp:55]: For future reference: it would have been nice to place the refactoring (which is where this function came about) in a separate commit. I believe this refactoring is a big improvement, but also it caused the biggest code churn here.

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp:84]: I know I should have commented on this days ago, but anyway: I like the idea, but would feel the `.value` field is a better place for that than the `.type`? It's not a property that is tied to the type, is it? I'm also saying this as the type continues to exist (that very same `irept` can be shared across all the instantiations), while the value will be overwritten anyway.

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp]: I'd suggest following `clang-format`'s suggestions throughout this patch as at least some of the changes do use the proposed formatting.

File: src/jbmc/jbmc_parse_options.cpp
Comment (tautschnig) [inline src/jbmc/jbmc_parse_options.cpp]: Nit pick: I don't think this comment adds value.

File: src/goto-programs/remove_returns.cpp
Comment (smowton) [reply src/goto-programs/remove_returns.cpp:84]: Neither is ideal -- really I want a `map<irep_idt, irept> attributes` defined for each symbol and GOTO function, but that's a breaking change. I chose the type because I thought it less likely to get copied into a place where the annotation no longer makes sense. I also wanted this to work with symbols with value `nil`, and nil-with-annotation-comment seemed dubious if anyone uses `full_eq` to compare them. However I can switch it if you prefer.

File: src/goto-programs/lazy_goto_model.h
Comment (smowton) [reply src/goto-programs/lazy_goto_model.h:31]: See new docs on the interface

**Comment by smowton** [comment]
@tautschnig @peterschrammel changes applied.

**Comment by smowton** [comment]
@tautschnig I also note that traviscistatus.com reports a 1800 job backlog for OSX nodes and 180 OSX nodes total, so I'm guessing it'll be at least half a day before anything runs.

File: src/goto-programs/remove_returns.cpp
Comment (tautschnig) [reply src/goto-programs/remove_returns.cpp:84]: Yes, one of the variants that are a breaking change is the right way to go. With that agreed, I'd think that `.value` is a better place for it isn't getting duplicated during symbolic execution, but for sure it's not the perfect place either. I'll leave it to you to decide, certainly not a blocker in any way.

File: src/goto-programs/lazy_goto_model.h
Comment (tautschnig) [reply src/goto-programs/lazy_goto_model.h:31]: Thanks a lot for the clarification!

**Review by tautschnig — APPROVED** [review]
I'm ok with this now, with one proposed change left at @smowton's discretion.

**Comment by smowton** [comment]
Confirmed with @peterschrammel in person that he is happy, so I'll merge this once Travis is done with this and the test-gen bump.

**Comment by tautschnig** [comment]
I'll start reviewing the PRs based off of that one once merged and rebased. I'd appreciate a ping once that's the case.

**Review by peterschrammel — APPROVED** [review]

**Comment by smowton** [comment]
Only OSX pending. Merging rather than wait for more merge conflicts as the OSX builds take days at the moment.

**Pull request merged**
